### ISSUE TITLE
Conv1D output changes

### ISSUE THREAD

**Issue by ThisIsIsaac** [pytorch/pytorch#23727]
## 🐛 Bug

Given the same input & weight (yes, we manually gave weight), and with `torch.backends.cudnn.deterministic = True` turned on, the output of 

```python
weight = # some code that reads weight file

conv = nn.Conv1D(...)
conv.weight.data = weight
```

changes evertime it is called:

**example outputs:**

1. iteration 1

```
[-3.0552e+00, -5.3343e+00, -6.5944e-01,  1.1911e+00,  4.3999e+00,
           1.4698e+00, -3.8650e+00,  1.4742e+00,  1.2590e+00,  5.3744e+00,
          -1.1283e+01,  1.1128e+01, -1.3646e+01,  1.2124e+00, -9.6420e-01,
           7.5311e+00, -5.4766e+00,  2.8123e+00, -9.1796e+00,  6.2736e+00, ...
```

2. iteration 2

```
-2.9771e+00, -5.2562e+00, -5.8136e-01,  1.2692e+00,  4.4779e+00,
           1.5479e+00, -3.7869e+00,  1.5522e+00,  1.3371e+00,  5.4525e+00,
          -1.1205e+01,  1.1206e+01, -1.3568e+01,  1.2905e+00, -8.8612e-01,
           7.6092e+00, -5.3985e+00,  2.8903e+00, -9.1016e+00,  6.3517e+00, ...
```

You can see that the outputs are different.

However, when we set the random seed with: `torch.manual_seed(0)`, then the output becomes identical on every iteration.

1. why does the output differ given the same inputs and weights, and with `torch.backends.cudnn.deterministic = True`?

2. Why does `torch.manual_seed(0)` make the outputs identical? 

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.0.0
 - OS (e.g., Linux): CentOS7
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): 
 - Python version: 3.5
 - CUDA/cuDNN version: 10.0 / 7
 - GPU models and configuration: V100

**Comment by xsacha** [comment]
Where did you set the input? Did you use dataloader?

**Comment by minty99** [comment]
Just these lines of code:

```
class ConvNorm(torch.nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size=1, stride=1,
                 padding=None, dilation=1, bias=True, w_init_gain='linear'):
        super(ConvNorm, self).__init__()
        if padding is None:
            assert(kernel_size % 2 == 1)
            padding = int(dilation * (kernel_size - 1) / 2)

        self.conv = torch.nn.Conv1d(in_channels, out_channels,
                                    kernel_size=kernel_size, stride=stride,
                                    padding=padding, dilation=dilation,
                                    bias=bias)

        # torch.nn.init.xavier_uniform_(
        #     self.conv.weight, gain=torch.nn.init.calculate_gain(w_init_gain))

    def forward(self, signal):
        conv_signal = self.conv(signal)
        return conv_signal

conv = ConvNorm(2, 32, 31)
conv.conv.weight.data = weight
output_tensor = conv(input_tensor
with open("test_conv_out.txt", "w") as f:
        f.write(str(output_tensor.size()) + "\n")
        f.write(str(output_tensor))
```

**Comment by vincentqb** [comment]
There are many places where non-deterministic behavior creeps in, as mentioned in the [documentation](https://pytorch.org/docs/stable/notes/randomness.html#pytorch) or in the [forum](https://discuss.pytorch.org/t/what-is-the-differenc-between-cudnn-deterministic-and-cudnn-benchmark/38054). You can use `torch.manual_seed()` to seed the RNG for all devices (both CPU and CUDA). Have you tried with `manual_seed` but not `torch.backends.cudnn.deterministic = True`? In any case, I expect more than one option to be required in order to get as deterministic a behavior as possible. 

Please use https://discuss.pytorch.org for questions. Please kindly feel encouraged to reopen this issue, if this does not apply.

**Comment by ThisIsIsaac** [comment]
@vincentqb 

> You can use torch.manual_seed() to seed the RNG for all devices (both CPU and CUDA).

We are aware of what  `torch.manual_seed()` but there should be no RNG factor in convolution. I've checked the cuDNN documentation on forward convolution and there is no parameter to take seed or any mention of RNG of convolution. 

> Have you tried with manual_seed but not torch.backends.cudnn.deterministic = True?

We've tried 2 settings: one with only `torch.backends.cudnn.deterministic = True` and another with both `torch.backends.cudnn.deterministic = True` and `manual_seed` set. Since convolution has no RNG factor, this shouldn't make any difference, but it seems to. Setting `manual_seed` fixes the results.
